### PR TITLE
fix pruning of block header db

### PR DESCRIPTION
**Pull request by larskuhtz**
The existing implementation of pruning could in rare cases result in a block header that was missing its payload. This could happen if all blocks in a prefix of a fork were marked as false positives but the respective payloads were not marked. (False positive blocks with an unmarked predecessor are detected and deleted.)

The PR fixes the issue by detecting and deleting any falsely marked block for which the payload hasn't been marked. This correct, because the payloads for all blocks on the root forks are marked.

This guarantees that for each block header the respective payload exists. It also lowers the false positive rate for block headers even further. 

Note, that we don't require that for each payload there is a respective block header. We therefor don't do any effort to detect falsely marked payloads.

### CODE REVIEW

File: src/Chainweb/BlockHeaderDB.hs [hunk 1]
```diff
@@ -324,7 +330,7 @@ pruneForks logger cdb callback limit = do
             -- Delete nodes which parent isn't in the filter or is in the
             -- falsePositiveSet
             parentIsMarked <- stToIO $ BF.elem p marked
-            if not (isGenesisBlockHeader h) && (not parentIsMarked || HS.member p falsePositiveSet)
+            if not (isGenesisBlockHeader h) && (not parentIsMarked || HS.member p falsePositiveSet || not isPayloadMarked)
```
Comment (fosskers) [inline src/Chainweb/BlockHeaderDB.hs]: Laziness question: if the `HS.member` call is true, will the `IO` involved in `isPayloadMarked` be ran? I'm trying to figure out which is better to place first.

Comment (larskuhtz) [reply src/Chainweb/BlockHeaderDB.hs]: Binding in `IO` is generally strict. Lazy `IO` is only used for "true" input/output functions. I would expected that `stToIO` evaluates its argument strictly.

Comment (larskuhtz) [reply src/Chainweb/BlockHeaderDB.hs]: I guess it depends on whether `BF.elem` is strict or returns a thunk. I would expect the former, but not sure.

Comment (fosskers) [reply src/Chainweb/BlockHeaderDB.hs]: In that case, it might be a tiny bit more performant to put the `isPayloadMarked` check before the `HS.member`.

Comment (larskuhtz) [reply src/Chainweb/BlockHeaderDB.hs]: I tried to refactor but my attempts led to more complex control flow. All of these checks should be pretty cheap (no IO involved), so I don't think optimizing this is worth the effort.

**Pull request merged**